pika/adapters/select_connection.py

```diff
@@ -217,120 +195,202 @@ def add_timeout(self, deadline, callback_method):
-        if error:
-            events |= ERROR
+    def poll(self, write_only=False):
+        """Wait for events on interested filedescriptors.
 
-        if events:
-            self._handler(self.fileno, events, write_only=write_only)
+        :param bool write_only: Passed through to the hadnlers to indicate
+            that they should only process write events.
+        """
 
-    def process_timeouts(self):
-        """Process the self._timeouts event stack"""
-        start_time = time.time()
-        # while loop instead of a more straightforward for loop so we can
-        # delete items from the list while iterating
-        i = 0
-        while i < len(self._timeouts):
-            t_id, timeout = self._timeouts[i]
-            if timeout['deadline'] <= start_time:
-                callback = timeout['callback']
-                del self._timeouts[i]
-                callback()
-            else:
-                i += 1
+        try:
+            read, write, error = select.select(self._fd_events[READ],
+                                               self._fd_events[WRITE],
+                                               self._fd_events[ERROR],
+                                               self.get_next_deadline())
+        except select.error as error:
+            if error.errno != errno.EINTR:
+                raise
 
-    def remove_timeout(self, timeout_id):
-        """Remove a timeout if it's still in the timeout stack
+        # Build an event bit mask for each fileno we've recieved an event for
 
-        :param str timeout_id: The timeout id to remove
+        fd_event_map = defaultdict(int)
+        for fd_set, ev in zip((read, write, error), (READ, WRITE, ERROR)):
+            for fileno in fd_set:
+                fd_event_map[fileno] |= ev
+
+        self._process_fd_events(fd_event_map, write_only)
+
+    def _process_fd_events(self, fd_event_map, write_only):
+        """ Processes the callbacks for each fileno we've recieved events.
+            Before doing so we re-calculate the event mask based on what is 
+            currently set in case it has been changed under our feet by a 
+            previous callback. We also take a store a refernce to the
+            fd_event_map in the class so that we can detect removal of an
+            fileno during processing of another callback and not generate
+            spurious callbacks on it.
 
+            :param dict fd_event_map: Map of fds to events recieved on them.
         """
-        for i in xrange(len(self._timeouts)):
-            t_id, timeout = self._timeouts[i]
-            if t_id == timeout_id:
-                del self._timeouts[i]
-                break
 
-    def start(self):
-        """Start the main poller loop. It will loop here until self.closed"""
-        while self.open:
-            self.poll()
-            self.process_timeouts()
-            self._manage_event_state()
+        self._processing_fd_event_map = fd_event_map
```

Comment:
So, as it stands, a read can trigger a write which could recursively drop into poll with the write_only flag set. When/if #556 is merged this nasty recursive behaviour goes away. Until then, whilst being ugly, does it cause a problem?

We'd end up overwriting self._processing_fd_event_map which would mean we'd drop out of the outer loop immediately after the inner. We'd then pop straight out of poll again with any events that hadn't been triggered. I guess if you had a fd that was always readable you could starve events on other descriptors?

Hopefully #556 will get merged shortly, if not I'll think about it some more.